### PR TITLE
Add Device Trust Requirement icon to web UI

### PR DESCRIPTION
**Pull request by avatus**
This PR expands upon the previous Device Trust topbar icon by adding a new icon to signal to users that device trust is required if they did not authorize the session

Screenshot 2024-08-01 at 4 23 52 PM

changelog: The web UI will now notify users of restricted access if their web session is not authorized by Device Trust (if required)

### CODE REVIEW

**Comment by avatus** [comment]
@codingllama i've decided to just recalculate the device trust requirements when we renew the session. Let me know if this seems ok to you

File: web/packages/design/src/Icon/Icons/ShieldWarning.tsx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* MIT License
```
Comment (codingllama) [inline web/packages/design/src/Icon/Icons/ShieldWarning.tsx:19]: Is the double-license here correct? Is it mandated by the codegen tool?

File: lib/auth/sessions.go [hunk 2]
```diff
@@ -363,6 +363,17 @@ func (a *Server) newWebSession(
 	if err != nil {
 		return nil, nil, trace.Wrap(err)
 	}
+
+	if tdr, err := a.calculateTrustedDeviceMode(ctx, func() ([]types.Role, error) {
+		return checker.Roles(), nil
+	}); err != nil {
+		log.
+			WithError(err).
+			Warnf("Failed to calculate trusted device mode for session")
```
Comment (codingllama) [inline lib/auth/sessions.go:372]: nit: Nothing to format.

```suggestion
			Warn("Failed to calculate trusted device mode for session")
```

File: lib/auth/sessions.go [hunk 3]
```diff
@@ -363,6 +363,17 @@ func (a *Server) newWebSession(
 	if err != nil {
 		return nil, nil, trace.Wrap(err)
 	}
+
+	if tdr, err := a.calculateTrustedDeviceMode(ctx, func() ([]types.Role, error) {
```
Comment (codingllama) [inline lib/auth/sessions.go:367]: If we are doing this on newWebSession then could we remove the other invocations in favor of this one?

File: lib/auth/sessions.go [hunk 4]
```diff
@@ -363,6 +363,17 @@ func (a *Server) newWebSession(
...
+		return checker.Roles(), nil
+	}); err != nil {
+		log.
+			WithError(err).
+			Warnf("Failed to calculate trusted device mode for session")
+	} else {
+		sess.SetTrustedDeviceRequirement(tdr)
```
Comment (codingllama) [inline lib/auth/sessions.go:374]: Note that this is now going to be persisted along with the session, so we should probably update the comments on the proto field.

**Review by codingllama — COMMENTED** [review]
Thanks!

File: web/packages/design/src/Icon/Icons/ShieldWarning.tsx (hunk 1 above)
Comment (avatus) [reply web/packages/design/src/Icon/Icons/ShieldWarning.tsx:19]: Yes this code was autogenerated. I believe it handles "our" license as well as the license to use the icon in this way? I'm not sure, but I'll leave it as the generator probably thought about it more than I did

File: lib/auth/sessions.go (hunk 3 above)
Comment (avatus) [reply lib/auth/sessions.go:367]: Yes! good catch

**Review by codingllama — APPROVED** [review]

**Comment by codingllama** [comment]
Issue #43126.

**Review by kimlisa — APPROVED** [review]

File: web/packages/teleport/src/Login/useLogin.ts [hunk 5]
```diff
@@ -123,6 +130,7 @@ type DeviceWebToken = {
 
 type LoginResponse = {
   deviceWebToken?: DeviceWebToken;
+  trusted_device_requirement?: TrustedDeviceRequirement;
```
[the remark this answers is not in the document]
Comment (codingllama) [reply web/packages/teleport/src/Login/useLogin.ts:133]: If you mean the json tag on WebSessionSpecV2.TrustedDeviceRequirement, then I would not change it. All other fields in there are lower_snake, so this one is just following suite.

[the remark this answers is not in the document]
Comment (avatus) [reply web/packages/teleport/src/Login/useLogin.ts:133]: I meant on `CreateSessionResponse`, where [all others are camelCase](https://github.com/gravitational/teleport/blob/782c18ae3678807ef525b65ef525047a12d6f865/lib/web/apiserver.go#L2158)

[the remark this answers is not in the document]
Comment (codingllama) [reply web/packages/teleport/src/Login/useLogin.ts:133]: This wasn't used until now, right? Go for it.

**Pull request merged**